htdocs/core/class/discount.class.php

```diff
@@ -407,6 +409,15 @@ function unlink_invoice()
         $resql = $this->db->query($sql);
         if ($resql)
         {
+            require_once DOL_DOCUMENT_ROOT.'/core/class/interfaces.class.php';
```

Comment:
We must try to limit triggers to CRUD event, so Create, Read, Update Delete and give context data into this->context() to be more accurate.
So here, you can use the DISCOUNT_MODIFY trigger name.
And to be able to make a difference with an simple update of a discount record, you can set, just before the call of trigger

$this->context=array('unlinkfrominvoiceid'=>$rowidinvoice);	    
just before the call ->run_triggers